### PR TITLE
[Buganizer ID: 544723192]Updated the logic to properly handle empty WHOIS fallback responses

### PR DESCRIPTION
**Pull request by jayapradha-p**
## [Buganizer ID: 544723192]Updated the logic to properly handle empty WHOIS fallback responses

---

## Description

Please provide a detailed description of your changes. This helps reviewers understand your work and its context.

**What problem does this PR solve?**
Updated the Whois action in the Enrichment integration to properly handle empty responses during classic WHOIS fallback (e.g., when Port 43 is restricted or server queries return empty contact dictionaries). Added a has_whois_data validation helper and warning logging in data_model.py and integrated it into Whois.py
 so that empty fallback payloads are treated as failures rather than false-positive enrichments, correctly setting result_value = "false" and outputting "No entities were enriched.". Comprehensive unit and integration test coverage was added in test_whois.py, and the integration version was bumped to 39.0.
 
 Case Scope Support: Implemented ExecutionScope detection and resolved target alert identifiers from the case's open alerts (open_alerts/alerts). This passes the resolved alert_identifier to CreateEntity during graph extension, eliminating the 404 "Not found alert with id in case" error in Case Scope.
 
**How does this PR solve the problem?**
(e.g., "Modified algorithm in `src/foo.js`," "Added new component `Bar.vue`," "Updated dependency `baz` to version 1.2.3.")

**Any other relevant information (e.g., design choices, tradeoffs, known issues):**
(e.g., "Chose approach A over B due to performance considerations," "This change might affect X in certain edge cases," "Requires manual migration steps for existing users.")

---

## Checklist:

Please ensure you have completed the following items before submitting your PR.
This helps us review your contribution faster and more efficiently.

### General Checks:

- [x] I have read and followed the project's [contributing.md](https://github.com/chronicle/marketplace/blob/main/docs/contributing.md) guide.
- [x] My code follows the project's coding style guidelines.
- [x] I have performed a self-review of my own code.
- [x] My changes do not introduce any new warnings.
- [x] My changes pass all existing tests.
- [x] I have added new tests where appropriate to cover my changes. (If applicable)
- [x] I have updated the documentation where necessary (e.g., README, API docs). (If applicable)

### Open-Source Specific Checks:

- [x] My changes do not introduce any Personally Identifiable Information (PII) or sensitive customer data.
- [x] My changes do not expose any internal-only code examples, configurations, or URLs.
- [x] All code examples, comments, and messages are generic and suitable for a public repository.
- [x] I understand that any internal context or sensitive details related to this work are handled separately in internal systems (Buganizer for Google team members).

### For Google Team Members and Reviewers Only:

- [x] I have included the Buganizer ID in the PR title or description (e.g., "Internal Buganizer ID: 123456789" or "Related Buganizer: go/buganizer/123456789").
- [x] I have ensured that all internal discussions and PII related to this work remain in Buganizer.
- [x] I have tagged the PR with one or more labels that reflect the pull request purpose.

---

## Screenshots (If Applicable)

If your changes involve UI or visual elements, please include screenshots or GIFs here.
Ensure any sensitive data is redacted or generalized.

---

## Further Comments / Questions

Any additional comments, questions, or areas where you'd like specific feedback.

### CODE REVIEW

File: content/response_integrations/power_ups/enrichment/actions/Whois.py [hunk 1]
```diff
@@ -126,6 +126,13 @@ def main():
                     domain = get_domain_from_string(entity.identifier)
                     if domain:
                         whois_data = get_domain_whois(domain, logger=siemplify.LOGGER)
+                        if not has_whois_data(whois_data):
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/actions/Whois.py]: `get_domain_whois` already validates the fetched data through `has_whois_data` across both the RDAP query and the classic WHOIS fallback paths, returning an empty dictionary `{}` whenever no meaningful data is present. Re-evaluating `has_whois_data(whois_data)` here is redundant and couples the action script to internal data model validation helpers. We can simplify this check to standard falsiness:
```suggestion
                        if not whois_data:
```
(and remove `has_whois_data` from the imports at line 35).

File: content/response_integrations/power_ups/enrichment/core/data_model.py [hunk 2]
```diff
@@ -408,6 +408,42 @@ def process_entities(entity_list: list[dict[str, Any]]) -> None:
...
+        if key == "contacts":
+            if isinstance(value, dict):
+                for contact in value.values():
+                    if contact is None:
+                        continue
+                    if isinstance(contact, dict):
+                        if any(bool(v) for v in contact.values()):
+                            return True
+                    elif hasattr(contact, "to_dict"):
+                        if bool(contact.to_dict()):
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/core/data_model.py]: There is an inconsistency here between dictionary contacts and `ContactInfo` instances. `ContactInfo.to_dict()` strips out `None` keys (`{k: v for k, v in self.__dict__.items() if v is not None}`), but preserves keys with empty strings (e.g., `{"name": ""}`). Because `bool({"name": ""})` evaluates to `True`, `ContactInfo(name="")` causes `has_whois_data` to return `True`, whereas a raw dictionary `{"name": ""}` correctly evaluates to `False` via line 432.

To ensure uniform behavior and prevent false-positive enrichment on empty string contact attributes, check the truthiness of the values returned by `to_dict()` as well:
```suggestion
                    elif hasattr(contact, "to_dict"):
                        if any(bool(v) for v in contact.to_dict().values()):
                            return True
```

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py [hunk 3]
```diff
@@ -468,3 +471,143 @@ def test_map_rdap_to_whois_deterministic_emails() -> None:
     result = map_rdap_to_whois(rdap_with_emails)
     assert result["emails"] == ["b@example.com", "a@example.com"]
 
+
+@pytest.mark.execution_scope("Alert")
+@set_metadata(
+    parameters={"Create Entities": "true", "Domain Age Threshold": "0"},
+    input_context={"environment": "Default", "alert_id": "alert_1"},
+)
+def test_whois_action_fallback_empty_response(
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:486]: The `script_session` fixture parameter is unused throughout this test function. Dropping it cleans up unused parameters and keeps the function within the 3-argument limit:
```suggestion
def test_whois_action_fallback_empty_response(
    product: EnrichmentProduct,
    action_output: MockActionOutput,
    mock_siemplify: MagicMock,
) -> None:
    """Test Whois action fallback handling when classic WHOIS query returns an empty response."""
```

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py [hunk 4]
```diff
@@ -468,3 +471,143 @@ def test_map_rdap_to_whois_deterministic_emails() -> None:
...
+
+    assert action_output.results.execution_state == ExecutionState.COMPLETED
+    end_msg, result_val, _ = mock_siemplify.end.call_args[0]
+    assert result_val == "false"
+    assert "No entities were enriched." in end_msg
+    assert "Failed to enrich the following entities: positivacaoseguros.com.br" in end_msg
+    res_list = mock_siemplify.result.add_result_json.call_args[0][0]
+    assert res_list == []
+    mock_siemplify.update_entities.assert_not_called()
+
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:534]: Add a docstring describing the test scenario, move the import to the top of the module, and consider adding test assertions for empty-string attributes on `ContactInfo` (e.g. `ContactInfo(name="")`) to guard against false-positive validation:
```suggestion
def test_has_whois_data_validation() -> None:
    """Validate has_whois_data heuristics against varied payloads, empty models, and edge cases."""
```

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py [hunk 5]
```diff
@@ -468,3 +471,143 @@ def test_map_rdap_to_whois_deterministic_emails() -> None:
...
+    assert has_whois_data({"contacts": {"registrant": ContactInfo(name="Test Org")}})
+
+
+def test_get_domain_whois_empty_fallback_logs_warning() -> None:
+    from ...core.data_model import get_domain_whois
+
+    mock_logger = MagicMock()
+    with patch("requests.get") as mock_get:
+        mock_get.return_value.status_code = 404
+
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:611]: Add a concise docstring explaining the test scenario and move the import to the top of the file:
```suggestion
def test_get_domain_whois_empty_fallback_logs_warning() -> None:
    """Test that get_domain_whois emits a warning and returns an empty dict on empty fallback."""
```

File: content/response_integrations/power_ups/enrichment/core/data_model.py [hunk 6]
```diff
@@ -408,6 +408,42 @@ def process_entities(entity_list: list[dict[str, Any]]) -> None:
     return whois_data.to_dict()
 
 
+def has_whois_data(whois_data: dict[str, Any] | None) -> bool:
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/core/data_model.py:435]: `has_whois_data` currently contains 12 conditional branches spanning 5 levels of diagonal nesting to inspect contacts and field values. We can dramatically flatten and simplify this by adhering to single level of abstraction (Rule G34) and encapsulating the contact inspection into a dedicated helper predicate:

```python
def _has_contact_data(contact: Any) -> bool:
    """Check whether a contact object or dictionary contains non-empty attributes."""
    if not contact:
        return False
    if hasattr(contact, "to_dict"):
        contact = contact.to_dict()
    if isinstance(contact, dict):
        return any(bool(val) for val in contact.values())
    return True


def _has_meaningful_field(key: str, value: Any) -> bool:
    """Check whether a specific WHOIS field contains non-empty, actionable data."""
    if key == "raw" or not value:
        return False
    if key == "contacts":
        if isinstance(value, dict):
            return any(_has_contact_data(c) for c in value.values())
        return bool(value)
    if isinstance(value, (list, tuple, set)):
        return any(bool(item) for item in value)
    return True


def has_whois_data(whois_data: dict[str, Any] | None) -> bool:
    \"\"\"Check whether a WHOIS/RDAP data dictionary contains any meaningful enrichment data.

    Args:
        whois_data: Dictionary of parsed WHOIS or RDAP fields.

    Returns:
        True if at least one meaningful field (excluding 'raw') contains data, False otherwise.
    \"\"\"
    if not whois_data or not isinstance(whois_data, dict):
        return False

    return any(_has_meaningful_field(key, val) for key, val in whois_data.items())
```

This decomposition:
1. Replaces 5 levels of diagonal indentation with flat, 1-level guard clauses and declarative `any(...)` generator expressions.
2. Unifies dict and `ContactInfo` object attribute truthiness checking, resolving the edge case where `ContactInfo(name="")` returned `True`.
3. Ensures each function has a single, cohesive responsibility.

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 3 above)
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:486]: Add a concise 1-line docstring describing the scenario and expected assertions:
```suggestion
def test_whois_action_fallback_empty_response(
    product: EnrichmentProduct,
    action_output: MockActionOutput,
    mock_siemplify: MagicMock,
) -> None:
    """Test Whois action fallback handling when classic WHOIS query returns an empty response."""
```

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py [hunk 7]
```diff
@@ -468,3 +471,143 @@ def test_map_rdap_to_whois_deterministic_emails() -> None:
...
+            Whois.main()
+
+    assert action_output.results.execution_state == ExecutionState.COMPLETED
+    end_msg, result_val, _ = mock_siemplify.end.call_args[0]
+    assert result_val == "false"
+    assert "No entities were enriched." in end_msg
+    assert "Failed to enrich the following entities: positivacaoseguros.com.br" in end_msg
+    res_list = mock_siemplify.result.add_result_json.call_args[0][0]
+    assert res_list == []
+    mock_siemplify.update_entities.assert_not_called()
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:533]: Add a concise 1-line docstring describing the validation suite:
```suggestion
def test_has_whois_data_validation() -> None:
    """Validate has_whois_data heuristics against varied payloads, empty models, and edge cases."""
```

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py [hunk 8]
```diff
@@ -468,3 +471,143 @@ def test_map_rdap_to_whois_deterministic_emails() -> None:
...
+    })
+    assert not has_whois_data({
+        "contacts": {
+            "registrant": ContactInfo(),
+        }
+    })
+
+    assert has_whois_data({"id": ["example.com"]})
+    assert has_whois_data({"registrar": ["MarkMonitor Inc."]})
+    assert has_whois_data({"contacts": {"registrant": {"name": "Test Org"}}})
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:602]: Add a concise 1-line docstring explaining the test behavior:
```suggestion
def test_get_domain_whois_empty_fallback_logs_warning() -> None:
    """Test that get_domain_whois emits a warning and returns an empty dictionary on empty fallback."""
```

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py [hunk 9]
```diff
@@ -103,6 +103,8 @@ def mock_extract(target, *args, **kwargs):
             mock_res.registered_domain = "univ-lyon1.fr"
         elif identifier == "failed.com":
             mock_res.registered_domain = "failed.com"
+        elif identifier == "positivacaoseguros.com.br":
```
Comment (adarshtwy) [inline content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:114]: Notice that `mock_extract` is missing a docstring as well as parameter and return type annotations:
```python
def mock_extract(target: Any, *args: Any, **kwargs: Any) -> MagicMock:
    """Mock tldextract.extract to simulate domain resolution for test targets."""
```

**Review by adarshtwy — CHANGES_REQUESTED** [review]

File: content/response_integrations/power_ups/enrichment/actions/Whois.py (hunk 1 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/actions/Whois.py]: Resolved

File: content/response_integrations/power_ups/enrichment/core/data_model.py (hunk 2 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/core/data_model.py]: Resolved

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 3 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:486]: Resolved

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 4 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:534]: Resolved

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 5 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:611]: Resolved

File: content/response_integrations/power_ups/enrichment/core/data_model.py (hunk 6 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/core/data_model.py:435]: Resolved

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 7 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:533]: Resolved

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 8 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:602]: Resolved

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 9 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:114]: Resolved

File: content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py (hunk 3 above)
Comment (jayapradha-p) [reply content/response_integrations/power_ups/enrichment/tests/test_actions/test_whois.py:486]: Resolved

**Review by ujdutt1 — DISMISSED** [review]
LGTM!
